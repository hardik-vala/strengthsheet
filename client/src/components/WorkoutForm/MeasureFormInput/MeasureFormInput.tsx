import { useState } from "react";
import { View } from "react-native";
import { TextInput } from "react-native-paper";
import {
  DrillSet,
  Exercise,
  ExerciseMeasure,
  WorkoutValueKey,
  isValidMeasureValue,
} from "../../../models/Workout/Core";
import { MeasureHistoryList } from "../MeasureHistoryList/MeasureHistoryList";
import { ExerciseMeasureHistoryRecord, WorkoutValues } from "../common";

export interface MeasureFormInputProps {
  circuitKey?: string;
  exercise: Exercise;
  set: DrillSet;
  measure: ExerciseMeasure;
  measureHistory: ExerciseMeasureHistoryRecord[];
  workoutValues: WorkoutValues;
  onUpdateWorkoutValues: (updatedWorkoutValues: WorkoutValues) => void;
}

export function MeasureFormInput({
  circuitKey,
  exercise,
  set,
  measure,
  measureHistory,
  workoutValues,
  onUpdateWorkoutValues,
}: MeasureFormInputProps) {
  const [currWorkoutValue, setCurrWorkoutValue] = useState("");
  const [isError, setIsError] = useState(false);

  const workoutValueKeyStr = WorkoutValueKey.create(
    circuitKey,
    exercise.key,
    set.index,
    set.setType,
    measure.key
  ).toString();

  function getPlaceholder() {
    if (!measureHistory || measureHistory.length == 0) {
      return "";
    }

    const sortedHistory = [...measureHistory];
    sortedHistory.sort(
      (r1, r2) => r2.timestamp.getTime() - r1.timestamp.getTime()
    );

    return sortedHistory[0].value;
  }

  return (
    <View>
      <TextInput
        defaultValue=""
        dense={true}
        error={isError}
        mode="flat"
        onChangeText={(text) => {
          const isValid = isValidMeasureValue(text, measure.unit);
          setCurrWorkoutValue(text);
          setIsError(text && !isValid);
          if (!text || isValid) {
            workoutValues[workoutValueKeyStr] = text;
            onUpdateWorkoutValues(workoutValues);
          }
        }}
        placeholder={getPlaceholder()}
        style={{
          textAlign: "right",
          height: 30,
          width: 80,
        }}
        value={currWorkoutValue}
      />
      {measureHistory && measureHistory.length > 0 && (
        <MeasureHistoryList measureHistory={measureHistory} />
      )}
    </View>
  );
}
