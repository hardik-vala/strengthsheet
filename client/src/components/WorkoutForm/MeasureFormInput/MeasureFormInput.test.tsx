import { act, fireEvent, render, screen } from "@testing-library/react-native";
import React from "react";
import {
  ExerciseUnit,
  SetType,
  WorkoutValueKey,
} from "../../../models/Workout/Core";
import { MeasureFormInput } from "./MeasureFormInput";

const TEST_MEASURE = {
  key: "test_measure",
  displayName: "reps",
  unit: ExerciseUnit.reps,
};

const TEST_EXERCISE = {
  key: "test_exercise",
  displayName: "Test Exercise",
  measures: [
    { key: "lb", displayName: "lb", unit: ExerciseUnit.lb },
    TEST_MEASURE,
  ],
};

const TEST_CIRCUIT = {
  key: "test_circuit",
  exercises: [{ exercise: TEST_EXERCISE, note: null }],
};

const TEST_SET = { index: 2, setType: SetType.Working };

const TEST_MEASURE_HISTORY = [
  {
    timestamp: new Date("2023-08-28"),
    key: WorkoutValueKey.createFromExercise(
      "test_exercise",
      2,
      SetType.Working,
      "test_measure"
    ),
    value: "13",
  },
  {
    timestamp: new Date("2023-08-27"),
    key: WorkoutValueKey.createFromExercise(
      "test_exercise",
      2,
      SetType.Working,
      "test_measure"
    ),
    value: "12",
  },
  {
    timestamp: new Date("2023-08-28"),
    key: WorkoutValueKey.createFromCircuit(
      "test_circuit",
      "test_exercise",
      2,
      SetType.Working,
      "test_measure"
    ),
    value: "23",
  },
  {
    timestamp: new Date("2023-08-27"),
    key: WorkoutValueKey.createFromCircuit(
      "test_circuit",
      "test_exercise",
      2,
      SetType.Working,
      "test_measure"
    ),
    value: "12",
  },
];

describe("MeasureFormInput", () => {
  it("renders correctly", () => {
    const onError = jest.fn();

    render(
      <MeasureFormInput
        exercise={TEST_EXERCISE}
        set={TEST_SET}
        measure={TEST_MEASURE}
        measureHistory={[]}
        workoutValues={{}}
        onUpdateWorkoutValues={() => {}}
        onError={onError}
      />
    );

    expect(onError).not.toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });

  it("renders with history", () => {
    const onError = jest.fn();

    render(
      <MeasureFormInput
        exercise={TEST_EXERCISE}
        set={TEST_SET}
        measure={TEST_MEASURE}
        measureHistory={TEST_MEASURE_HISTORY}
        workoutValues={{}}
        onUpdateWorkoutValues={() => {}}
        onError={onError}
      />
    );

    expect(onError).not.toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });

  it("calls onUpdateWorkoutValues when text input changes", () => {
    const onUpdateWorkoutValues = jest.fn();
    const onError = jest.fn();

    render(
      <MeasureFormInput
        exercise={TEST_EXERCISE}
        set={TEST_SET}
        measure={TEST_MEASURE}
        measureHistory={TEST_MEASURE_HISTORY}
        workoutValues={{}}
        onUpdateWorkoutValues={onUpdateWorkoutValues}
        onError={onError}
      />
    );

    act(() => {
      fireEvent.changeText(screen.getByPlaceholderText("13"), "8");
    });

    expect(onUpdateWorkoutValues).toHaveBeenCalledWith(
      expect.objectContaining({
        "test_exercise:2:3:test_measure": "8",
      })
    );
    expect(onError).not.toHaveBeenCalled();
  });

  it("renders circuit measure inputs correctly", () => {
    const onError = jest.fn();

    render(
      <MeasureFormInput
        exercise={TEST_EXERCISE}
        set={TEST_SET}
        measure={TEST_MEASURE}
        circuitKey={TEST_CIRCUIT.key}
        measureHistory={TEST_MEASURE_HISTORY}
        workoutValues={{}}
        onUpdateWorkoutValues={() => {}}
        onError={onError}
      />
    );

    expect(onError).not.toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });

  describe("when the text input is invalid", () => {
    it("renders the error indicator", () => {
      const onError = jest.fn();

      render(
        <MeasureFormInput
          exercise={TEST_EXERCISE}
          set={TEST_SET}
          measure={TEST_MEASURE}
          measureHistory={TEST_MEASURE_HISTORY}
          workoutValues={{}}
          onUpdateWorkoutValues={() => {}}
          onError={onError}
        />
      );

      act(() => {
        fireEvent.changeText(screen.getByPlaceholderText("13"), "Q");
      });

      expect(onError).toHaveBeenCalled();
      expect(screen).toMatchSnapshot();
    });

    it("doesn't call onUpdateWorkoutValues", () => {
      const onUpdateWorkoutValues = jest.fn();
      const onError = jest.fn();

      render(
        <MeasureFormInput
          exercise={TEST_EXERCISE}
          set={TEST_SET}
          measure={TEST_MEASURE}
          measureHistory={TEST_MEASURE_HISTORY}
          workoutValues={{}}
          onUpdateWorkoutValues={onUpdateWorkoutValues}
          onError={onError}
        />
      );

      act(() => {
        fireEvent.changeText(screen.getByPlaceholderText("13"), "Q");
      });

      expect(onUpdateWorkoutValues).not.toHaveBeenCalled();
      expect(onError).toHaveBeenCalled();
    });
  });
});
