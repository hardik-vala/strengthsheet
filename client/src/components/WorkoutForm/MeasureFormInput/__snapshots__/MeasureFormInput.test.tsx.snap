// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasureFormInput renders circuit measure inputs correctly 1`] = `
<View>
  <View
    style={
      {
        "marginRight": 15,
        "width": 75,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          {
            "backgroundColor": "rgba(232, 222, 248, 1)",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderWidth": 1,
          },
          {
            "borderRadius": 5,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 0,
            },
            {
              "minHeight": 30,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(98, 91, 113, 1)"
          defaultValue=""
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="13"
          placeholderTextColor="darkgray"
          selectionColor="rgba(98, 91, 113, 1)"
          style={
            [
              {
                "margin": 0,
                "zIndex": 1,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 7,
                "paddingTop": 7,
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              {
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`MeasureFormInput renders correctly 1`] = `
<View>
  <View
    style={
      {
        "marginRight": 15,
        "width": 75,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          {
            "backgroundColor": "rgba(232, 222, 248, 1)",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderWidth": 1,
          },
          {
            "borderRadius": 5,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 0,
            },
            {
              "minHeight": 30,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(98, 91, 113, 1)"
          defaultValue=""
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="darkgray"
          selectionColor="rgba(98, 91, 113, 1)"
          style={
            [
              {
                "margin": 0,
                "zIndex": 1,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 7,
                "paddingTop": 7,
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              {
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`MeasureFormInput renders with history 1`] = `
<View>
  <View
    style={
      {
        "marginRight": 15,
        "width": 75,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          {
            "backgroundColor": "rgba(232, 222, 248, 1)",
            "borderColor": "transparent",
            "borderRadius": 4,
            "borderWidth": 1,
          },
          {
            "borderRadius": 5,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 0,
            },
            {
              "minHeight": 30,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(98, 91, 113, 1)"
          defaultValue=""
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="13"
          placeholderTextColor="darkgray"
          selectionColor="rgba(98, 91, 113, 1)"
          style={
            [
              {
                "margin": 0,
                "zIndex": 1,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 7,
                "paddingTop": 7,
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              {
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`MeasureFormInput when the text input is invalid renders the error indicator 1`] = `
<View>
  <View
    style={
      {
        "marginRight": 15,
        "width": 75,
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          {
            "backgroundColor": "rgba(232, 222, 248, 1)",
            "borderColor": "rgba(179, 38, 30, 1)",
            "borderRadius": 4,
            "borderWidth": 2,
          },
          {
            "borderRadius": 5,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 0,
            },
            {
              "minHeight": 30,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          cursorColor="rgba(179, 38, 30, 1)"
          defaultValue=""
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="13"
          placeholderTextColor="darkgray"
          selectionColor="rgba(179, 38, 30, 1)"
          style={
            [
              {
                "margin": 0,
                "zIndex": 1,
              },
              {
                "height": 30,
              },
              {
                "paddingBottom": 7,
                "paddingTop": 7,
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "paddingHorizontal": 16,
                "textAlign": "center",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              {
                "fontSize": 12,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value="Q"
        />
      </View>
    </View>
  </View>
</View>
`;
